### ISSUE TITLE
Consumer.cpp may calculate RTP timestamp offset wrongly

### ISSUE THREAD

**Issue by ibc** [versatica/mediasoup#301]
Currently all the `XxxxxConsumer.cpp` classes rewrite the RTP packet `timestamp` as follows:

```c++
// Calculate RTP timestamp diff between now and last sent RTP packet.
if (this->rtpStream->GetMaxPacketMs() != 0u)
{
	auto now    = DepLibUV::GetTime();
	auto diffMs = now - this->rtpStream->GetMaxPacketMs();
	auto diffTs = diffMs * this->rtpStream->GetClockRate() / 1000;

	this->rtpTimestampManager.Offset(diffTs);
}
```

As @ggarber pointed out "if a packet arrives with 200ms delay, it will leave the SFU with a timestamp 200ms higher than it should". This is:

* The browser sends packet1 (contains a complete video frame) with ts = 1000 (assume ts is ms).
* It takes 50ms for packet1 to reach the SFU.
* The SFU rewrites its ts to (let's say) 2500.
* The browser sends packet2 (contains a complete video frame) with ts = 1100.
* It takes 250ms for packet2 to reach the SFU.
* The SFU rewrites its ts to 2800.

This is wrong since, based on those ts values, the consumer endpoint will render packet2 300ms after rendering packet1.

Instead of using `now` we may have to use the latest SenderReport to get the sender's wallclock and use it to correctly apply our timestamp offset.

### RTCP Sender Report

> **NTP timestamp: 64 bits**
>
> Indicates the wallclock time when this report was sent so that it may be used in combination with timestamps returned in reception reports from other receivers to measure round-trip propagation to those receivers. Receivers should expect that the measurement accuracy of the timestamp may be limited to far less than the resolution of the NTP timestamp. The measurement uncertainty of the timestamp is not indicated as it may not be known. A sender that can keep track of elapsed time but has no notion of wallclock time may use the elapsed time since joining the session instead. This is assumed to be less than 68 years, so the high bit will be zero. It is permissible to use the sampling clock to estimate elapsed wallclock time. A sender that has no notion of wallclock or elapsed time may set the NTP timestamp to zero.
>
> **RTP timestamp: 32 bits**
>
> Corresponds to the same time as the NTP timestamp (above), but in the same units and with the same random offset as the RTP timestamps in data packets. This correspondence may be used for intra- and inter-media synchronization for sources whose NTP timestamps are synchronized, and may be used by media- independent receivers to estimate the nominal RTP clock frequency. Note that in most cases this timestamp will not be equal to the RTP timestamp in any adjacent data packet. Rather, it is calculated from the corresponding NTP timestamp using the relationship between the RTP timestamp counter and real time as maintained by periodically checking the wallclock time at a sampling instant.

**Comment by ibc** [comment]
In our `RtpStreamRecv.cpp` we are reading the Sender Report NTP timestamp upon receipt of a Sender Report:

```c++
void RtpStreamRecv::ReceiveRtcpSenderReport(RTC::RTCP::SenderReport* report)
{
	MS_TRACE();

	this->lastSrReceived  = DepLibUV::GetTime();
	this->lastSrTimestamp = report->GetNtpSec() << 16;
	this->lastSrTimestamp += report->GetNtpFrac() >> 16;

	// Update the score with the current RR.
	UpdateScore();
}
```

However we just use it when generating Receiver Reports. Probably the `RtpStreamRecv` should expose an API to retrieve the latest NTP timestamp so the `Consumer` uses it to properly calculate the packet RTP timestamp.

**Comment by ibc** [comment]
mmmm, I may be wrong. mediasoup just recalculates the  `diffTs` when receiving a "sync packet" (this is, a key frame packet that is used to enable forwarding of a different simulcast stream, or after pausing the consumer, etc):

```c++
// Sync sequence number and timestamp if required.
if (isSyncPacket)
{
	if (packet->IsKeyFrame())
		MS_DEBUG_TAG(rtp, "sync key frame received");

	this->rtpSeqManager.Sync(packet->GetSequenceNumber());
	this->rtpTimestampManager.Sync(packet->GetTimestamp());

	// Calculate RTP timestamp diff between now and last sent RTP packet.
	if (this->rtpStream->GetMaxPacketMs() != 0u)
	{
		auto now    = DepLibUV::GetTime();
		auto diffMs = now - this->rtpStream->GetMaxPacketMs();
		auto diffTs = diffMs * this->rtpStream->GetClockRate() / 1000;

		this->rtpTimestampManager.Offset(diffTs);
	}

	this->syncRequired = false;
}
```

So in the example above (in the issue description), since packet2 is not a "sync packet", the ts offset that will be applied to it does not depend on "now" at all.

**Comment by ibc** [comment]
The only possible issue coming to my mind my happen if the "sync packet" arrives with delay (imagine it becomes it was lost and comes again in a retransmission). If so, we'll honor `now` and, from now on, all packets will be forwarded with such a "delayed offset".

> Given the Sender Report, we get a mapping between RTP timestamp and clock timestamp per ssrc. So when wishing to do stream switching, we can convert it to clock timestamp to calculate diffTs, and divide it by clock rate to convert it to diffRtpTs. Or something like that.

**Comment by ibc** [comment]
The importat thing here is avoid lipsync problems between audio and video from the same producer. AFAIU, the Sender Reports of both, audio and video, have something in common: the `NTP timestamp` (the `RTP timestamp` in SR "corresponds to the same time as the NTP timestamp but in the same units and with the same random offset as the RTP timestamps in data packets".

So, what I understand is that, somehow, we should take the SR NPM timestamp as reference to calculate our custom ts offsets when switching simulcast streams. No idea what this means, though.

**Comment by jmillan** [comment]
>As @ggarber pointed out "if a packet arrives with 200ms delay, it will leave the SFU with a timestamp 200ms higher than it should". This is:

>* The browser sends packet1 (contains a complete video frame) with ts = 1000 (assume ts is ms).
>* It takes 50ms for packet1 to reach the SFU.
>* The SFU rewrites its ts to (let's say) 2500.
>* The browser sends packet2 (contains a complete video frame) with ts = 1100.
>* It takes 250ms for packet2 to reach the SFU.
>* The SFU rewrites its ts to 2800.

>This is wrong since, based on those ts values, the consumer endpoint will render packet2 300ms after rendering packet1.

Good point,

Indeed we can take the `maxPacketMs` from the ts of the incoming RTP packet instead of taking it from the system time. This will take the encoder clock as the reference, exactly what is needed.

**Comment by jmillan** [comment]
> Indeed we can take the maxPacketMs from the ts of the incoming RTP packet instead of taking it from the system time. This will take the encoder clock as the reference, exactly what is needed.

Or simply store the original RTP ts, no need to convert it to ms.

**Comment by ibc** [comment]
WIP: https://github.com/versatica/mediasoup/commit/b3d1e43abc2a3a2c1e3c448269df02781a3d1886?ts=2 and https://github.com/versatica/mediasoup/commit/65375cbc024874a36640863b39c4d10dd6b1e066?ts=2

**Comment by ibc** [comment]
@jmillan please check if this is correct: https://github.com/versatica/mediasoup/commit/65375cbc024874a36640863b39c4d10dd6b1e066?ts=2#diff-df9d53eb39e095cb8126208e7fb93408R430

**Comment by ibc** [comment]
Done in 586e31cde4471703e5fd28a59fea30b8d36a6b85 and published in 3.0.2.
